File: bytedance/string/003. Longest Substring Without Repeating Characters.go
//Time Complexity: O(N)

func lengthOfLongestSubstring(s string) int {
    m := map[byte]int{}
    n := len(s)
    r, ans := -1, 0
    for i := 0; i < n; i++ {
        if i != 0 {
            delete(m, s[i - 1])
        }
        for r + 1 < n && m[s[r + 1]] == 0 {
            m[s[r + 1]]++
            r++
        }
        ans = max(ans, r - i + 1)
    }
    return ans
}

func max(x, y int) int {
    if x < y {
        return y
    }
    return x
}

File: bytedance/string/003. Longest Substring Without Repeating Characters.py
# Time Complexity: O(N)
# Python
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        o = set()
        ans, r = 0, -1
        n = len(s)
        for i in range(n):
            if i != 0: 
                o.remove(s[i - 1])
            while r + 1 < n and s[r + 1] not in o:
                o.add(s[r + 1])
                r += 1
            ans = max(ans, r - i + 1)
        return ans

